packages/edit-navigation/src/components/menu-locations-editor/use-menu-locations.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * External dependencies
+ */
+import { map, flatMap, forEach, filter, groupBy } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { useDispatch } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
+
+export default function useMenuLocations() {
+	const { saveMenu } = useDispatch( 'core' );
+	const [ menuLocations, setMenuLocations ] = useState( null );
+
+	// a local state which maps menus to locations
+	// so that we can send one call per menu when
+	// updating locations, otherwise, without this local state
+	// we'd send one call per location
+	const [ menuLocationMap, setMenuLocationMap ] = useState( {} );
+
+	const fetchmenuLocations = async () => {
+		const path = '/__experimental/menu-locations';
+		const apiLocations = await apiFetch( {
+			path,
+			method: 'GET',
+		} );
+		return flatMap( apiLocations );
+	};
+
+	const initmenuLocations = async () => {
+		const latestLocations = await fetchmenuLocations();
+		setMenuLocations( latestLocations );
+	};
+
+	// we need to fecth the list of locations
+	// because the menu location entity
+	// caches their menu associations
+	useEffect( () => {
+		initmenuLocations();
+	}, [] );
+
+	// as soon as we have the menus we group
+	// all locations by the menuId they are assigned to
+	useEffect( () => {
+		if ( menuLocations ) {
+			const locationsByMenu = groupBy( menuLocations, 'menu' );
+			forEach( locationsByMenu, ( location, menuId ) => {
+				locationsByMenu[ menuId ] = map( location, 'name' );
+			} );
```

Comment:
```suggestion
			const locationsByMenu = mapValues(
				groupBy( menuLocations, 'menu' ),
				( locations ) => map( locations, 'name' )
			);
```

Reply:
It seems a little bit cleaner without the forEach and overriding keys. Also requires importing mapValues from lodash.